rpy2_arrow/pyarrow_rarrow.py

```diff
@@ -191,28 +191,37 @@ def rarrow_to_py_field(
     return pyarrow.lib.Field._import_from_c(_pyarrow_ptr(schema_ptr))
 
 
-def pyarrow_to_r_table(
-        obj: 'pyarrow.lib.Table',
-        py2rpy: typing.Optional[
-            conversion.Converter] = None
-):
+_as_arrow_table_from_stream_ptr = rinterface.evalr(
```

Comment:
I know that this is how I originally suggested doing it, but I was curious for completeness to see if it was faster to avoid getting a pointer to an R function handle and just directly passing the pointer into the source code using a format string. The overhead was of order 1ms; so not huge, but also relatively around 10% of the time spent converting. I'm fine either way.

i.e.
```
def _as_r_arrow_table_from_stream_ptr(reader_ptr):
    return rinterface.evalr(f"arrow::as_arrow_table(arrow::RecordBatchReader$import_from_c({str(_pyarrow_ptr(reader_ptr))})")
```

Reply:
And contrary to the above, there's also the possibility of just directly reusing `pyarrow_to_r_recordbatchreader` from above. You'd need to pass around function pointers, but then you'd reduce the duplicate code bits.

Reply:
> I know that this is how I originally suggested doing it, but I was curious for completeness to see if it was faster to avoid getting a pointer to an R function handle and just directly passing the pointer into the source code using a format string. The overhead was of order 1ms; so not huge, but also relatively around 10% of the time spent converting. I'm fine either way.

It is actually slower on this end:

```python
rinterface = pyra.rinterface
_pyarrow_ptr = pyra._pyarrow_ptr
def _pyarrow_table_to_r_table_ri(tbl: pyarrow.lib.Table) -> rinterface.SexpEnvironment:
    """Create an R `arrow` Table to mirror an Arrow Table in `pyarrow`."""
    reader_ptr = pyra.ffi.new('struct ArrowArrayStream*')
    tbl.to_reader()._export_to_c(_pyarrow_ptr(reader_ptr))
    return rinterface.evalr(
        f"arrow::as_arrow_table(arrow::RecordBatchReader$import_from_c({str(_pyarrow_ptr(reader_ptr))}))"
    )

orig = pyra._pyarrow_table_to_r_table_ri
pyra._pyarrow_table_to_r_table_ri = _pyarrow_table_to_r_table_ri
```
```python
%%timeit
with rpy2.robjects.conversion.localconverter(conv) as cv:
    cv.py2rpy(tbl)
```
```
3.28 ms ± 1.21 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

I think that the reason is that in the implementation I proposed the R code in a string is parsed only once at import time rather than each time an evaluation is made, and the parsing is more expensive than calling an R function.

Reply:
> And contrary to the above, there's also the possibility of just directly reusing `pyarrow_to_r_recordbatchreader` from above. You'd need to pass around function pointers, but then you'd reduce the duplicate code bits.

Excellent idea. There is a slight drop in performances though (see below). The amount of duplicated code seems minimal. I'll add a code comment to indicate that there is a sligth duplication in the name of efficiency. Unless you point out that there is better way.

```python
_as_arrow_table_from_stream_ptr = rinterface.evalr(
    """
    function(recordbatchreader) {
      arrow::as_arrow_table(recordbatchreader)
    }
    """
)

def _pyarrow_table_to_r_table_ri(tbl: pyarrow.lib.Table) -> rinterface.SexpEnvironment:
    """Create an R `arrow` Table to mirror an Arrow Table in `pyarrow`."""
    recordbatchreader = pyra.pyarrow_to_r_recordbatchreader(tbl.to_reader())
    return _as_arrow_table_from_stream_ptr(recordbatchreader)

def pyarrow_table_to_r_table(obj: pyarrow.lib.Table):
    """Create an R `arrow::Table` object from a pyarrow Table.

    This is sharing the C/C++ object between the two languages.
    The returned object depends on the active conversion rule in
    rpy2. By default it will be an `rpy2.robjects.Environment`.

    :param:obj: A :class:`pyarrow.lib.Table`.
    :return: The result of conversion rules for R `arrow` `Table` objects.
    By default a :class:`rpy2.robjects.Environment`.
    """
    # `res` is a low-level (rinterface-level) rpy2 object. This is an
    # rpy2 robject-level function. Use the conversion.
    res = _pyarrow_table_to_r_table_ri(obj)
    return pyra.conversion.get_conversion().py2rpy(res)

conv.py2rpy.register(pyarrow.lib.Table, pyarrow_table_to_r_table)
```
```python
%%timeit
with rpy2.robjects.conversion.localconverter(conv) as cv:
    cv.py2rpy(tbl)
```
```
3.34 ms ± 1.2 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

Reply:
> I think that the reason is that in the implementation I proposed the R code in a string is parsed only once at import time rather than each time an evaluation is made, and the parsing is more expensive than calling an R function.

Ah yes... I was testing my own code and redefining the function every time. I can confirm it is faster once you the function is taken out of the loop. Good catch.


> Excellent idea. There is a slight drop in performances though (see below). The amount of duplicated code seems minimal. I'll add a code comment to indicate that there is a sligth duplication in the name of efficiency. Unless you point out that there is better way.

I saw dips of ~0.1 - 0.5 ms. I wasn't able to find any improvements. As far as I'm concerned, that's within the margin of error and we can choose whichever approach is cleanest. Given the small amount of code duplication, it could go either way, and so I think your approach is good.

Reply:
You are right that the performance difference might be too small to count. I have removed the code duplication as you suggest.